drivers/network/meson/ethernet.c

```diff
@@ -210,18 +210,24 @@ static void tx_return(void)
 static void handle_irq()
 {
     uint32_t e = eth_dma->status;
-    if (e & DMA_INTR_RXF) {
-        rx_return();
-    }
-    if (e & DMA_INTR_TXF) {
-        tx_return();
-    }
-    if (e & DMA_INTR_ABNORMAL) {
-        if (e & DMA_INTR_FBE) {
-            sddf_dprintf("Ethernet device fatal bus error\n");
+    eth_dma->status &= e;
```

Comment:
Can you explain why we write back the value of `eth_dma->status` immediately after reading it? Is this to acknowledge the IRQs? If so, why do we do it before we've actually handled the IRQ? For performance?

Reply:
This is not the same thing as an acknowledgement of IRQs. When we initialise the driver IRQ mechanism, we tell the device which events we want to receive IRQs for (enabling interrupts). We select a subset from the set of event which can be configured to generate IRQs.

When these events occur (whether they generate IRQs or not), the corresponding bit is set to 1 in the status register. If we have enabled that event to send an IRQ, this will also generate an IRQ. If a second event occurs before the IRQ is handled, then it's bit will also be set to 1 in the status register.

When we eventually handle the IRQ, we read the status register and loop through all the bits. To re-enable IRQs, we both need to clear the status register (indicating that we are ready for another IRQ from this event) and ack the IRQ.

If while handling an IRQ, you read the status register and find out a packet has been received and a packet has been transmitted. You handle the RX event, then go on to handle the TX event. Then you write back to the status register to clear the RX and TX bits.

But, if while you were handling the TX event, a __new__ packet comes in, than there would be no change in the status register since the RX bit is still set. Then when you finished handling the TX event and write your RX and TX bits to the status register, you are actually clearing the RX bit that was originally set for the first packet, but has in the meantime been "reset" for the second packet. So you miss out on handling the secondary RX event.

If you had cleared the register immediately, then when you re-checked the register again after handling the event, you would see that the RX bit has been reset since you last checked.

This is my understanding anyway. So I think it's actually a correctness thing? But to be honest I'm not entirely sure... But in either case, I don't think that writing to the status register is the same thing as indicating that the event has been handled, I think its more like re-enabling IRQs generated by this event.

Reply:
This does seem to be the case from reading the HW manual:

Interrupt Event Register
"When an event occurs that sets a bit in EIR, an interrupt occurs if the corresponding bit in the interrupt mask register (EIMR) is also set. Writing a 1 to an EIR bit clears it; writing 0 has no effect. This register is cleared upon hardware reset."

Interrupt Mask Register
"EIMR controls which interrupt events are allowed to generate actual interrupts. A hardware reset clears this register. If the corresponding bits in the EIR and EIMR registers are set, an interrupt is generated. The interrupt signal remains asserted until a 1 is written to the EIR field (write 1 to clear) or a 0 is written to the EIMR field."